File: tools/o3d2opaque.py
#!/usr/bin/env python3
"""Build the host-side opaque-texture qualification sidecar.

One output byte belongs to one O3D source triangle.  A value of one means a
normal (non-semitransparent) textured packet may use the prepared 16-bit CLUT
and omit the per-pixel PS1-palette-word validity test.  Zero is always safe:
false negatives cost speed only, while a false positive would fill a real
texture hole and is therefore forbidden.

The native TIM page is authoritative.  Pages with no zero palette word among
their referenced indices qualify immediately.  For a holed page, the tool
constructs an integer-only conservative footprint:

* every closed unit texel cell intersected by the affine UV triangle, and
* a two-texel Chebyshev dilation, with 8-bit texture-coordinate wrapping.

The closed-cell test covers the floor operation at an exact edge.  The
dilation is the current 240x224/Q8.8 rasterizer's conservative cost-model
allowance for chain and span-gradient truncation plus the sub-pixel prestep.
It is intentionally larger than the old floating-point M_uv scan.  The
companion tools/opaque_selftest.py gate then checks every available recorded
rasterizer packet directly: no qualified source triangle may report even one
bit-17 rejection.  Enlarging the render target or changing the fixed-point
walk requires regenerating and revalidating this sidecar.
"""

import argparse
import os
import struct
import sys


O3D_HEADER_BYTES = 24
O3D_POINT_BYTES = 12
O3D_NORMAL_BYTES = 12
O3D_POLYGON_WORDS = 16

TIM_CLUT_DATA_OFFSET = 20
TIM_CLUT_ENTRIES = 256
TIM_PIXEL_DATA_OFFSET = 544
TIM_SIZE = TIM_PIXEL_DATA_OFFSET + 256 * 256
TEXTURE_PAGE_IDS = (10, 12, 14, 26, 28, 30)
TPAGE_MASK = 0x1F
FOOTPRINT_DILATION = 2


class QualificationError(Exception):
    pass


def resolve_tpage(raw):
    """Resolve a packet TPage exactly like rasterize_packet."""
    page = raw & TPAGE_MASK
    return page if page in TEXTURE_PAGE_IDS else TEXTURE_PAGE_IDS[0]


def _cross(a, b, c):
    return ((b[0] - a[0]) * (c[1] - a[1])
            - (b[1] - a[1]) * (c[0] - a[0]))


def _point_on_segment(a, b, point):
    return (_cross(a, b, point) == 0
            and min(a[0], b[0]) <= point[0] <= max(a[0], b[0])
            and min(a[1], b[1]) <= point[1] <= max(a[1], b[1]))


def _segments_intersect(a, b, c, d):
    ab_c = _cross(a, b, c)
    ab_d = _cross(a, b, d)
    cd_a = _cross(c, d, a)
    cd_b = _cross(c, d, b)
    if ((ab_c == 0 and _point_on_segment(a, b, c))
            or (ab_d == 0 and _point_on_segment(a, b, d))
            or (cd_a == 0 and _point_on_segment(c, d, a))
            or (cd_b == 0 and _point_on_segment(c, d, b))):
        return True
    return ((ab_c > 0) != (ab_d > 0)
            and (cd_a > 0) != (cd_b > 0))


def _point_in_triangle(point, triangle):
    signs = [_cross(triangle[i], triangle[(i + 1) % 3], point)
             for i in range(3)]
    return all(value >= 0 for value in signs) or all(value <= 0 for value in signs)


def _cell_intersects_triangle(u, v, uvs):
    """Exact closed-cell/triangle intersection using doubled integers."""
    triangle = [(2 * x, 2 * y) for x, y in uvs]
    x0 = 2 * u
    y0 = 2 * v
    square = ((x0, y0), (x0 + 2, y0),
              (x0 + 2, y0 + 2), (x0, y0 + 2))

    if any(x0 <= x <= x0 + 2 and y0 <= y <= y0 + 2
           for x, y in triangle):
        return True
    if any(_point_in_triangle(point, triangle) for point in square):
        return True
    return any(_segments_intersect(triangle[i], triangle[(i + 1) % 3],
                                   square[j], square[(j + 1) % 4])
               for i in range(3) for j in range(4))


def conservative_affine_texels(uvs, dilation=FOOTPRINT_DILATION):
    """Return a conservative set of wrapped (u, v) texels for one triangle."""
    us = [u for u, _v in uvs]
    vs = [v for _u, v in uvs]
    footprint = set()
    # The -1 lower bound is needed because the closed cell immediately below
    # an integer vertex touches it.  It is conservative at a floor boundary.
    for v in range(min(vs) - 1, max(vs) + 1):
        for u in range(min(us) - 1, max(us) + 1):
            if _cell_intersects_triangle(u, v, uvs):
                footprint.add((u & 0xFF, v & 0xFF))

    for _unused in range(dilation):
        footprint = {((u + du) & 0xFF, (v + dv) & 0xFF)
                     for u, v in footprint
                     for du in (-1, 0, 1) for dv in (-1, 0, 1)}
    return footprint


def scan_texture_pages(texture_dir):
    pages = {}
    for page in TEXTURE_PAGE_IDS:
        path = os.path.join(texture_dir, "trex_texture_page_%d.tim" % page)
        try:
            with open(path, "rb") as handle:
                data = handle.read()
        except OSError as exc:
            raise QualificationError("cannot read %s: %s" % (path, exc))
        if len(data) < TIM_SIZE:
            raise QualificationError("%s is %d bytes, expected at least %d"
                                     % (path, len(data), TIM_SIZE))
        clut = struct.unpack_from("<%dH" % TIM_CLUT_ENTRIES,
                                  data, TIM_CLUT_DATA_OFFSET)
        pixels = data[TIM_PIXEL_DATA_OFFSET:TIM_SIZE]
        used = frozenset(pixels)
        invalid = frozenset(index for index in used if clut[index] == 0)
        pages[page] = {
            "pixels": pixels,
            "invalid_indices": invalid,
            "opaque": not invalid,
            "used_indices": len(used),
            "invalid_texels": sum(1 for index in pixels if index in invalid),
        }
    return pages


def read_o3d_polygons(path):
    try:
        with open(path, "rb") as handle:
            data = handle.read()
    except OSError as exc:
        raise QualificationError("cannot read %s: %s" % (path, exc))
    if len(data) < O3D_HEADER_BYTES:
        raise QualificationError("%s is shorter than the O3D header" % path)
    # Point/normal counts come from the O3D's own header (big-endian u16 at
    # offset 0/2), the same way every JS tool in this pipeline reads them
    # (see o3d2facenormals.js), rather than from constants that silently go
    # stale if the mesh's vertex or normal count ever changes.
    point_count, normal_count = struct.unpack_from(">HH", data, 0)
    offset = (O3D_HEADER_BYTES + point_count * O3D_POINT_BYTES
              + normal_count * O3D_NORMAL_BYTES)
    stride = O3D_POLYGON_WORDS * 2
    if len(data) < offset or (len(data) - offset) % stride:
        raise QualificationError("%s has a malformed polygon block" % path)
    return [struct.unpack_from(">%dH" % O3D_POLYGON_WORDS,
                               data, offset + index * stride)
            for index in range((len(data) - offset) // stride)]


def qualify_o3d(o3d_path, texture_dir, dilation=FOOTPRINT_DILATION):
    pages = scan_texture_pages(texture_dir)
    polygons = read_o3d_polygons(o3d_path)
    sidecar = bytearray(len(polygons))
    tally = {page: [0, 0] for page in TEXTURE_PAGE_IDS}
    untextured = 0

    for index, word in enumerate(polygons):
        if word[15] == 0:
            untextured += 1
            continue
        page = resolve_tpage(word[15])
        tally[page][0] += 1
        info = pages[page]
        qualifies = info["opaque"]
        if not qualifies:
            uvs = ((word[8] & 0xFF, word[9] & 0xFF),
                   (word[10] & 0xFF, word[11] & 0xFF),
                   (word[12] & 0xFF, word[13] & 0xFF))
            qualifies = not any(
                info["pixels"][v * 256 + u] in info["invalid_indices"]
                for u, v in conservative_affine_texels(uvs, dilation))
        if qualifies:
            sidecar[index] = 1
            tally[page][1] += 1
    return sidecar, tally, untextured, pages


def write_sidecar(path, payload):
    # Normal generation runs through make; keep the write atomic so an
    # interrupted qualification cannot leave a plausible truncated table.
    temporary = path + ".tmp"
    with open(temporary, "wb") as handle:
        handle.write(payload)
    os.replace(temporary, path)


def main(argv=None):
    parser = argparse.ArgumentParser()
    parser.add_argument("o3d")
    parser.add_argument("texture_dir")
    parser.add_argument("output")
    parser.add_argument("--dilation", type=int, default=FOOTPRINT_DILATION)
    args = parser.parse_args(argv)
    if args.dilation < 0:
        parser.error("--dilation must be non-negative")
    try:
        sidecar, tally, untextured, pages = qualify_o3d(
            args.o3d, args.texture_dir, args.dilation)
        write_sidecar(args.output, sidecar)
    except (OSError, QualificationError) as exc:
        print("opaque qualification failed: %s" % exc, file=sys.stderr)
        return 1

    print("%s: %d/%d textured triangles qualified; %d untextured stay on the flat path"
          % (args.output, sum(sidecar), len(sidecar) - untextured, untextured))
    for page in TEXTURE_PAGE_IDS:
        count, qualified = tally[page]
        info = pages[page]
        print("  page %2d: %4d/%4d triangles, %5d invalid texels, %3d used indices"
              % (page, qualified, count, info["invalid_texels"],
                 info["used_indices"]))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: tools/o3d2opaque.js
#!/usr/bin/env node
// Build the host-side opaque-texture qualification sidecar.
//
// One output byte belongs to one O3D source triangle.  A value of one means a
// normal textured packet may use the prepared 16-bit CLUT and omit the
// per-pixel PS1-palette-word validity test.  Zero is always safe: false
// negatives cost speed only, while a false positive would fill a real texture
// hole and is therefore forbidden.

const fs = require("fs");
const path = require("path");

const O3D_HEADER_BYTES = 24;
const O3D_POINT_BYTES = 12;
const O3D_NORMAL_BYTES = 12;
const O3D_POLYGON_WORDS = 16;

const TIM_CLUT_DATA_OFFSET = 20;
const TIM_CLUT_ENTRIES = 256;
const TIM_PIXEL_DATA_OFFSET = 544;
const TIM_SIZE = TIM_PIXEL_DATA_OFFSET + 256 * 256;
const TEXTURE_PAGE_IDS = [10, 12, 14, 26, 28, 30];
const TPAGE_MASK = 0x1f;
const FOOTPRINT_DILATION = 2;

function fail(message) {
    console.error("opaque qualification failed: " + message);
    process.exitCode = 1;
}

function resolveTpage(raw) {
    const page = raw & TPAGE_MASK;
    return TEXTURE_PAGE_IDS.includes(page) ? page : TEXTURE_PAGE_IDS[0];
}

function cross(a, b, c) {
    return ((b[0] - a[0]) * (c[1] - a[1])
        - (b[1] - a[1]) * (c[0] - a[0]));
}

function pointOnSegment(a, b, point) {
    return cross(a, b, point) === 0
        && Math.min(a[0], b[0]) <= point[0]
        && point[0] <= Math.max(a[0], b[0])
        && Math.min(a[1], b[1]) <= point[1]
        && point[1] <= Math.max(a[1], b[1]);
}

function segmentsIntersect(a, b, c, d) {
    const abC = cross(a, b, c);
    const abD = cross(a, b, d);
    const cdA = cross(c, d, a);
    const cdB = cross(c, d, b);

    if ((abC === 0 && pointOnSegment(a, b, c))
        || (abD === 0 && pointOnSegment(a, b, d))
        || (cdA === 0 && pointOnSegment(c, d, a))
        || (cdB === 0 && pointOnSegment(c, d, b))) {
        return true;
    }

    return (abC > 0) !== (abD > 0)
        && (cdA > 0) !== (cdB > 0);
}

function pointInTriangle(point, triangle) {
    const signs = triangle.map((_, index) =>
        cross(triangle[index], triangle[(index + 1) % 3], point));
    return signs.every(value => value >= 0)
        || signs.every(value => value <= 0);
}

function cellIntersectsTriangle(u, v, uvs) {
    // Exact closed-cell/triangle intersection using doubled integers.
    const triangle = uvs.map(([x, y]) => [2 * x, 2 * y]);
    const x0 = 2 * u;
    const y0 = 2 * v;
    const square = [[x0, y0], [x0 + 2, y0],
        [x0 + 2, y0 + 2], [x0, y0 + 2]];

    if (triangle.some(([x, y]) =>
        x0 <= x && x <= x0 + 2 && y0 <= y && y <= y0 + 2)) {
        return true;
    }
    if (square.some(point => pointInTriangle(point, triangle))) {
        return true;
    }

    return triangle.some((_, i) =>
        square.some((_, j) =>
            segmentsIntersect(triangle[i], triangle[(i + 1) % 3],
                square[j], square[(j + 1) % 4])));
}

function conservativeAffineTexels(uvs, dilation = FOOTPRINT_DILATION) {
    // Return a conservative set of wrapped (u, v) texels for one triangle.
    const us = uvs.map(([u]) => u);
    const vs = uvs.map(([, v]) => v);
    let footprint = new Set();

    // The -1 lower bound is needed because the closed cell immediately below
    // an integer vertex touches it.  It is conservative at a floor boundary.
    for (let v = Math.min(...vs) - 1; v <= Math.max(...vs) + 1; v++) {
        for (let u = Math.min(...us) - 1; u <= Math.max(...us) + 1; u++) {
            if (cellIntersectsTriangle(u, v, uvs)) {
                footprint.add(`${u & 0xff},${v & 0xff}`);
            }
        }
    }

    for (let iteration = 0; iteration < dilation; iteration++) {
        const expanded = new Set();
        for (const key of footprint) {
            const [u, v] = key.split(",").map(Number);
            for (const du of [-1, 0, 1]) {
                for (const dv of [-1, 0, 1]) {
                    expanded.add(`${(u + du) & 0xff},${(v + dv) & 0xff}`);
                }
            }
        }
        footprint = expanded;
    }

    return Array.from(footprint, key => key.split(",").map(Number));
}

function scanTexturePages(textureDir) {
    const pages = new Map();

    for (const page of TEXTURE_PAGE_IDS) {
        const texturePath = path.join(textureDir, `trex_texture_page_${page}.tim`);
        let data;
        try {
            data = fs.readFileSync(texturePath);
        } catch (error) {
            throw new Error(`cannot read ${texturePath}: ${error.message}`);
        }

        if (data.length < TIM_SIZE) {
            throw new Error(`${texturePath} is ${data.length} bytes, expected at least ${TIM_SIZE}`);
        }

        const clut = Array.from({ length: TIM_CLUT_ENTRIES }, (_, index) =>
            data.readUInt16LE(TIM_CLUT_DATA_OFFSET + index * 2));
        const pixels = data.subarray(TIM_PIXEL_DATA_OFFSET, TIM_SIZE);
        const used = new Set(pixels);
        const invalidIndices = new Set(
            Array.from(used).filter(index => clut[index] === 0));
        let invalidTexels = 0;

        for (const index of pixels) {
            if (invalidIndices.has(index)) {
                invalidTexels++;
            }
        }

        pages.set(page, {
            pixels,
            invalidIndices,
            opaque: invalidIndices.size === 0,
            usedIndices: used.size,
            invalidTexels
        });
    }

    return pages;
}

function readO3dPolygons(filePath) {
    let data;
    try {
        data = fs.readFileSync(filePath);
    } catch (error) {
        throw new Error(`cannot read ${filePath}: ${error.message}`);
    }

    if (data.length < O3D_HEADER_BYTES) {
        throw new Error(`${filePath} is shorter than the O3D header`);
    }

    // Point/normal counts come from the O3D header, the same way the other JS
    // tools read them, rather than from constants that can silently go stale.
    const pointCount = data.readUInt16BE(0);
    const normalCount = data.readUInt16BE(2);
    const offset = O3D_HEADER_BYTES
        + pointCount * O3D_POINT_BYTES
        + normalCount * O3D_NORMAL_BYTES;
    const stride = O3D_POLYGON_WORDS * 2;

    if (data.length < offset || (data.length - offset) % stride !== 0) {
        throw new Error(`${filePath} has a malformed polygon block`);
    }

    const polygons = [];
    for (let cursor = offset; cursor < data.length; cursor += stride) {
        const words = [];
        for (let index = 0; index < O3D_POLYGON_WORDS; index++) {
            words.push(data.readUInt16BE(cursor + index * 2));
        }
        polygons.push(words);
    }
    return polygons;
}

function qualifyO3d(o3dPath, textureDir, dilation = FOOTPRINT_DILATION) {
    const pages = scanTexturePages(textureDir);
    const polygons = readO3dPolygons(o3dPath);
    const sidecar = Buffer.alloc(polygons.length);
    const tally = new Map(TEXTURE_PAGE_IDS.map(page => [page, [0, 0]]));
    let untextured = 0;

    polygons.forEach((word, index) => {
        if (word[15] === 0) {
            untextured++;
            return;
        }

        const page = resolveTpage(word[15]);
        const pageTally = tally.get(page);
        pageTally[0]++;
        const info = pages.get(page);
        let qualifies = info.opaque;

        if (!qualifies) {
            const uvs = [[word[8] & 0xff, word[9] & 0xff],
                [word[10] & 0xff, word[11] & 0xff],
                [word[12] & 0xff, word[13] & 0xff]];
            qualifies = conservativeAffineTexels(uvs, dilation).every(([u, v]) =>
                !info.invalidIndices.has(info.pixels[v * 256 + u]));
        }

        if (qualifies) {
            sidecar[index] = 1;
            pageTally[1]++;
        }
    });

    return { sidecar, tally, untextured, pages };
}

function writeSidecar(filePath, payload) {
    // Normal generation runs through make; keep the write atomic so an
    // interrupted qualification cannot leave a plausible truncated table.
    const temporary = filePath + ".tmp";
    fs.writeFileSync(temporary, payload);
    fs.renameSync(temporary, filePath);
}

function parseArguments(argv) {
    if (argv.length < 3) {
        throw new Error("usage: o3d2opaque.js <o3d> <texture_dir> <output> [--dilation N]");
    }

    let dilation = FOOTPRINT_DILATION;
    if (argv.length > 3) {
        if (argv.length !== 5 || argv[3] !== "--dilation") {
            throw new Error("usage: o3d2opaque.js <o3d> <texture_dir> <output> [--dilation N]");
        }
        dilation = Number(argv[4]);
    }
    if (!Number.isInteger(dilation) || dilation < 0) {
        throw new Error("--dilation must be non-negative");
    }

    return { o3d: argv[0], textureDir: argv[1], output: argv[2], dilation };
}

try {
    const args = parseArguments(process.argv.slice(2));
    const result = qualifyO3d(args.o3d, args.textureDir, args.dilation);
    writeSidecar(args.output, result.sidecar);

    const qualified = Array.from(result.sidecar).reduce((sum, value) => sum + value, 0);
    console.log(`${args.output}: ${qualified}/${result.sidecar.length - result.untextured}`
        + ` textured triangles qualified; ${result.untextured} untextured stay on the flat path`);

    for (const page of TEXTURE_PAGE_IDS) {
        const [count, pageQualified] = result.tally.get(page);
        const info = result.pages.get(page);
        console.log(`  page ${String(page).padStart(2)}: ${String(pageQualified).padStart(4)}`
            + `/${String(count).padStart(4)} triangles, ${String(info.invalidTexels).padStart(5)}`
            + ` invalid texels, ${String(info.usedIndices).padStart(3)} used indices`);
    }
} catch (error) {
    fail(error.message);
}
